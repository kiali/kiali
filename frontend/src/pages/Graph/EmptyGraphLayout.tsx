import * as React from 'react';
import {
  Button,
  ButtonVariant,
  EmptyState,
  EmptyStateBody,
  EmptyStateIcon,
  EmptyStateVariant,
  EmptyStateHeader,
  EmptyStateFooter
} from '@patternfly/react-core';
import { kialiStyle } from 'styles/StyleUtils';
import * as _ from 'lodash';
import { Namespace } from '../../types/Namespace';
import { KialiIcon } from '../../config/KialiIcon';
import { DecoratedGraphElements } from '../../types/Graph';
import { RefreshIntervalManual } from 'config/Config';
import { IntervalInMilliseconds } from 'types/Common';
import { t } from 'utils/I18nUtils';

type EmptyGraphLayoutProps = {
  action?: any;
  elements?: DecoratedGraphElements;
  error?: string;
  isError: boolean;
  isLoading?: boolean;
  isMiniGraph: boolean;
  loaded: boolean;
  namespaces?: Namespace[];
  refreshInterval?: IntervalInMilliseconds;
  showIdleNodes?: boolean;
  toggleIdleNodes?: () => void;
};

const emptyStateStyle = kialiStyle({
  height: '98%',
  marginRight: 'auto',
  marginLeft: 'auto',
  marginBottom: 10,
  marginTop: 10
});

type EmptyGraphLayoutState = {};

export class EmptyGraphLayout extends React.Component<EmptyGraphLayoutProps, EmptyGraphLayoutState> {
  shouldComponentUpdate(nextProps: EmptyGraphLayoutProps): boolean {
    const currentIsEmpty = this.props.elements === undefined || _.isEmpty(this.props.elements.nodes);
    const nextIsEmpty = nextProps.elements === undefined || _.isEmpty(nextProps.elements.nodes);

    // Update if we have elements and we are not loading
    if (!nextProps.isLoading && !nextIsEmpty) {
      return true;
    }

    // Update if we are going from having no elements to having elements or vice versa
    if (currentIsEmpty !== nextIsEmpty) {
      return true;
    }
    // Do not update if we have elements and the namespace didn't change, as this means we are refreshing
    return !(!nextIsEmpty && _.isEqual(this.props.namespaces, nextProps.namespaces));
  }

  namespacesText(): React.ReactElement | null {
    if (this.props.namespaces && this.props.namespaces.length > 0) {
      if (this.props.namespaces.length === 1) {
        return (
          <div>
            <b>{this.props.namespaces[0].name}</b>
          </div>
        );
      } else {
        const namespacesString = `${this.props.namespaces
          .slice(0, -1)
          .map(namespace => namespace.name)
          .join(',')} and ${this.props.namespaces[this.props.namespaces.length - 1].name}`;
        return (
          <div>
            <b>{namespacesString}</b>
          </div>
        );
      }
    }
    return null;
  }

  render(): React.ReactNode {
    if (this.props.isError) {
      return (
        <EmptyState id="empty-graph-error" variant={EmptyStateVariant.lg} className={emptyStateStyle}>
          <EmptyStateHeader
            titleText={t('Error loading graph')}
            icon={<EmptyStateIcon icon={KialiIcon.Error} />}
            headingLevel="h5"
          />
          <EmptyStateBody>{this.props.error}</EmptyStateBody>
        </EmptyState>
      );
    }

    if (this.props.isLoading) {
      return (
        <EmptyState id="empty-graph-is-loading" variant={EmptyStateVariant.lg} className={emptyStateStyle}>
          <EmptyStateHeader titleText={t('Loading graph')} headingLevel="h5" />
        </EmptyState>
      );
    }

    const isGraphEmpty = !this.props.elements || !this.props.elements.nodes || this.props.elements.nodes.length < 1;

    if (this.props.namespaces?.length === 0) {
      return (
        <EmptyState id="empty-graph-no-namespace" variant={EmptyStateVariant.lg} className={emptyStateStyle}>
          <EmptyStateHeader titleText={t('No namespace is selected')} headingLevel="h5" />
          <EmptyStateBody>
            {t('There is currently no namespace selected, please select one using the Namespace selector.')}
          </EmptyStateBody>
        </EmptyState>
      );
    }

    if (this.props.refreshInterval === RefreshIntervalManual && !this.props.loaded && !this.props.isMiniGraph) {
      return (
        <EmptyState
          id="empty-graph-manual"
          data-test="manual-refresh"
          variant={EmptyStateVariant.lg}
          className={emptyStateStyle}
        >
          <EmptyStateHeader titleText={t('Manual refresh required')} headingLevel="h5" />
          <EmptyStateBody>
            {t(
              'The refresh interval is set to "Manual". To render the graph, select your desired filters and options and then click the Refresh button. Or, if preferred, change the setting to the desired interval.'
            )}
          </EmptyStateBody>
        </EmptyState>
      );
    }

    if (isGraphEmpty && !this.props.isMiniGraph) {
      return (
        <EmptyState id="empty-graph" variant={EmptyStateVariant.lg} className={emptyStateStyle}>
          <EmptyStateHeader titleText={t('Empty Graph')} headingLevel="h5" />
          <EmptyStateBody>
            {t(
              'There is currently no graph available for the selected namespaces. This usually means the namespaces do not have measurable traffic for the selected time period:'
            )}
            {this.namespacesText()}
            <br />
            {this.props.showIdleNodes && (
              <>
                {t(
                  'You are currently displaying "Idle nodes". Send requests to the service mesh and click the Refresh button.'
                )}
              </>
            )}
            {!this.props.showIdleNodes && (
              <>{t('You can enable "Idle Nodes" to display nodes that have yet to see any request traffic.')}</>
            )}
            <p>
              <br />
              {t('To read the FAQ entry about why your graph may be empty, ')}
              <a href={'https://kiali.io/docs/faq/graph/#emptygraph'} target="_blank" rel="noreferrer">
                {t('click here.')}
              </a>
            </p>
          </EmptyStateBody>
          <EmptyStateFooter>
            <Button
              onClick={this.props.showIdleNodes ? this.props.action : this.props.toggleIdleNodes}
              variant={ButtonVariant.primary}
            >
              {(this.props.showIdleNodes && <>{t('Refresh')}</>) || <>{t('Display idle nodes')}</>}
            </Button>
          </EmptyStateFooter>
        </EmptyState>
      );
    }

    if (isGraphEmpty && this.props.isMiniGraph) {
      return (
        <EmptyState id="empty-mini-graph" variant={EmptyStateVariant.lg} className={emptyStateStyle}>
          <EmptyStateHeader titleText={t('Empty Graph')} headingLevel="h5" />
          <EmptyStateBody>{t('No graph traffic for the time period.')}</EmptyStateBody>
        </EmptyState>
      );
    }

    return this.props.children;
  }
}
