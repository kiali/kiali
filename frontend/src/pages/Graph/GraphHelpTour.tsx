import { PopoverPosition } from '@patternfly/react-core';
import { TourStopInfo, TourInfo } from 'components/Tour/TourStop';
import { GraphShortcuts } from './GraphToolbar/GraphShortcuts';
import { t } from 'utils/I18nUtils';
import { GraphShortcutsPF } from './GraphToolbar/GraphShortcutsPF';

export const GraphTourStops: { [name: string]: TourStopInfo } = {
  ContextualMenu: {
    name: t('Contextual Menu'),
    description: t(
      'Right-click a node or an edge to see the contextual menu with links to details, traffic and inbound/outbound metrics for the node or edge.'
    ),
    position: PopoverPosition.left,
    distance: 250
  },
  ContextualMenuPF: {
    name: t('Contextual Menu'),
    description: t(
      'Click kebab menu on a node label for links to details, traffic, inbound/outbound metrics and node graph.'
    ),
    position: PopoverPosition.left,
    distance: 250
  },
  Display: {
    name: t('Display'),
    description: t(
      'Set edge labeling, node badging, and various display options. Response-time edge labeling, security badging, and traffic animation may affect performance. Response-times reflect the 95th percentile.'
    ),
    position: PopoverPosition.rightStart
  },
  Find: {
    name: t('Find and Hide'),
    description: t(
      'Highlight or Hide graph elements via typed expressions. Click the dropdown for preset Find or Hide expressions. Click the Find/Hide help icon for details on the expression language.'
    ),
    position: PopoverPosition.bottom
  },
  Graph: {
    name: t('Graph'),
    description:
      "Click on a node or edge to see its summary and emphasize its end-to-end paths. Double-click a node to see a graph focused on that node.\nDouble-click an t('external namespace' node to navigate directly to the namespace in the node's text label. Shift-Drag to quickly zoom in.",
    position: PopoverPosition.left,
    distance: 250
  },
  GraphPF: {
    name: t('Graph'),
    description: t('Click on a node or edge to see its summary and emphasize its end-to-end paths.'),
    position: PopoverPosition.left,
    distance: 250
  },
  GraphTraffic: {
    name: t('Graph Traffic'),
    description: t(
      'Choose the traffic rates used to generate the graph. Each supported protocol offers one or more options. Unused protocols can be omitted.'
    ),
    position: PopoverPosition.bottom
  },
  GraphType: {
    name: t('Graph Type'),
    description: t(
      'Select a workload, service or application graph view. An application view can optionally be versioned and relies on app and version labeling. Workload and service graphs provide physical and logical views, respectively.'
    ),
    position: PopoverPosition.right
  },
  Layout: {
    name: t('Layout selection'),
    description: t(
      'Select the graph layout for the mesh. Different layouts work best with different meshes. Find the layout that works best. Other buttons here provide zoom and fit-to-screen options.'
    ),
    position: PopoverPosition.right
  },
  Legend: {
    name: t('Legend'),
    description: t('Display the legend to learn about what the different shapes, colors and backgrounds mean.'),
    position: PopoverPosition.rightEnd
  },
  Namespaces: {
    name: t('Namespaces'),
    description: t('Select the namespaces you want to see in the graph.'),
    position: PopoverPosition.bottomStart
  },
  Shortcuts: {
    name: t('Shortcuts'),
    htmlDescription: GraphShortcuts,
    position: PopoverPosition.leftStart
  },
  ShortcutsPF: {
    name: t('Shortcuts'),
    htmlDescription: GraphShortcutsPF,
    position: PopoverPosition.leftStart
  },
  SidePanel: {
    name: t('Side Panel'),
    description: t(
      'The Side Panel shows details about the currently selected node or edge, otherwise the whole graph.'
    ),
    position: PopoverPosition.left
  },
  TimeRange: {
    name: t('Time Range & Replay'),
    description: t(
      'Select how often to refresh the graph and how much historical metric data is used to build the graph, per refresh. For example "Last 5m" means use the most recent 5 minutes of request metric data.  To replay a historical time window click the replay icon.  This replaces the current time range controls with the replay toolbar.'
    ),
    position: PopoverPosition.bottomEnd
  }
};

export const GraphTour: TourInfo = {
  name: t('GraphTour'),
  stops: [
    GraphTourStops.Shortcuts,
    GraphTourStops.Namespaces,
    GraphTourStops.GraphTraffic,
    GraphTourStops.GraphType,
    GraphTourStops.Display,
    GraphTourStops.Find,
    GraphTourStops.TimeRange,
    GraphTourStops.Graph,
    GraphTourStops.ContextualMenu,
    GraphTourStops.SidePanel,
    GraphTourStops.Layout,
    GraphTourStops.Legend
  ]
};

export const GraphTourPF: TourInfo = {
  name: t('GraphTour'),
  stops: [
    GraphTourStops.ShortcutsPF,
    GraphTourStops.Namespaces,
    GraphTourStops.GraphTraffic,
    GraphTourStops.GraphType,
    GraphTourStops.Display,
    GraphTourStops.Find,
    GraphTourStops.TimeRange,
    GraphTourStops.GraphPF,
    GraphTourStops.ContextualMenuPF,
    GraphTourStops.SidePanel,
    GraphTourStops.Layout,
    GraphTourStops.Legend
  ]
};
