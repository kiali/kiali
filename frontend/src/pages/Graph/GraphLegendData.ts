// Node Shapes
import workloadImage from '../../assets/img/graph/cy/node.svg';
import appImage from '../../assets/img/graph/cy/app.svg';
import serviceImage from '../../assets/img/graph/cy/service.svg';
import operationImage from '../../assets/img/graph/cy/operation.svg';
import serviceEntryImage from '../../assets/img/graph/cy/service-entry.svg';
// Node Colors
import nodeColorNormalImage from '../../assets/img/graph/cy/node-color-normal.svg';
import nodeColorWarningImage from '../../assets/img/graph/cy/node-color-warning.svg';
import nodeColorDangerImage from '../../assets/img/graph/cy/node-color-danger.svg';
import nodeColorIdleImage from '../../assets/img/graph/cy/node-color-idle.svg';
// Node Background
import externalNamespaceImage from '../../assets/img/graph/external-namespace.svg';
import restrictedNamespaceImage from '../../assets/img/graph/restricted-namespace.svg';
// Edges
import edgeSuccessImage from '../../assets/img/graph/cy/edge-success.svg';
import edgeDangerImage from '../../assets/img/graph/cy/edge-danger.svg';
import edgeWarnImage from '../../assets/img/graph/cy/edge-warn.svg';
import edgeIdlemage from '../../assets/img/graph/cy/edge-idle.svg';
import edgeTcpImage from '../../assets/img/graph/cy/edge-tcp.svg';
import edgeMtlsImage from '../../assets/img/graph/mtls-badge.svg';
// Traffic Animation
import trafficNormalImage from '../../assets/img/graph/cy/traffic-normal-request.svg';
import trafficFailedImage from '../../assets/img/graph/cy/traffic-failed-request.svg';
import trafficTcpImage from '../../assets/img/graph/cy/traffic-tcp.svg';
// Badges
import badgeCircuitBreakerImage from '../../assets/img/graph/node-badge-circuit-breaker.svg';
import badgeFaultInjectionImage from '../../assets/img/graph/node-badge-fault-injection.svg';
import badgeGatewaysImage from '../../assets/img/graph/node-badge-gateways.svg';
import badgeMirroringImage from '../../assets/img/graph/node-badge-mirroring.svg';
import badgeMissingSidecarImage from '../../assets/img/graph/node-badge-missing-sidecar.svg';
import badgeRequestTimeoutImage from '../../assets/img/graph/node-badge-request-timeout.svg';
import badgeTrafficShiftingSourceImage from '../../assets/img/graph/node-badge-traffic-shifting.svg';
import badgeTrafficSourceImage from '../../assets/img/graph/node-badge-traffic-source.svg';
import badgeVirtualServicesImage from '../../assets/img/graph/node-badge-virtual-services.svg';
import badgeWorkloadEntryImage from '../../assets/img/graph/node-badge-workload-entry.svg';
import badgeWaypointImage from '../../assets/img/graph/node-badge-waypoint.svg';
import { t } from 'utils/I18nUtils';
import { serverConfig } from '../../config';

export interface GraphLegendItem {
  data: GraphLegendItemRow[];
  isBadge?: boolean;
  title: string;
}

export interface GraphLegendItemRow {
  icon: string;
  label: string;
}

export const legendData = (): GraphLegendItem[] => {
  const nodesBadges = [
    { label: t('Circuit Breaker'), icon: badgeCircuitBreakerImage },
    { label: t('Fault Injection'), icon: badgeFaultInjectionImage },
    { label: t('Gateway'), icon: badgeGatewaysImage },
    { label: t('Mirroring'), icon: badgeMirroringImage },
    { label: t('Missing Sidecar'), icon: badgeMissingSidecarImage },
    { label: t('Request Timeout'), icon: badgeRequestTimeoutImage },
    { label: t('Traffic Shifting / TCP Traffic Shifting'), icon: badgeTrafficShiftingSourceImage },
    { label: t('Traffic Source'), icon: badgeTrafficSourceImage },
    { label: t('Virtual Service / Request Routing'), icon: badgeVirtualServicesImage },
    { label: t('Workload Entry'), icon: badgeWorkloadEntryImage }
  ];

  if (serverConfig.ambientEnabled) {
    nodesBadges.push({ label: t('Waypoint'), icon: badgeWaypointImage });
  }

  return [
    {
      title: t('Node Shapes'),
      data: [
        { label: t('Workload'), icon: workloadImage },
        { label: t('App'), icon: appImage },
        { label: t('Operation'), icon: operationImage },
        { label: t('Service'), icon: serviceImage },
        { label: t('Service Entry'), icon: serviceEntryImage }
      ]
    },
    {
      title: t('Node Colors'),
      data: [
        { label: t('Normal'), icon: nodeColorNormalImage },
        { label: t('Warn'), icon: nodeColorWarningImage },
        { label: t('Unhealthy'), icon: nodeColorDangerImage },
        { label: t('Idle'), icon: nodeColorIdleImage }
      ]
    },
    {
      title: t('Node Background'),
      data: [
        { label: t('Unselected Namespace'), icon: externalNamespaceImage },
        { label: t('Restricted / External'), icon: restrictedNamespaceImage }
      ]
    },
    {
      title: t('Edges'),
      data: [
        { label: t('Failure'), icon: edgeDangerImage },
        { label: t('Degraded'), icon: edgeWarnImage },
        { label: t('Healthy'), icon: edgeSuccessImage },
        { label: t('TCP Connection'), icon: edgeTcpImage },
        { label: t('Idle'), icon: edgeIdlemage },
        { label: t('mTLS (badge)'), icon: edgeMtlsImage }
      ]
    },
    {
      title: t('Traffic Animation'),
      data: [
        { label: t('Normal Request'), icon: trafficNormalImage },
        { label: t('Failed Request'), icon: trafficFailedImage },
        { label: t('TCP Traffic'), icon: trafficTcpImage }
      ]
    },
    {
      title: t('Node Badges'),
      isBadge: true,
      data: nodesBadges
    }
  ];
};
