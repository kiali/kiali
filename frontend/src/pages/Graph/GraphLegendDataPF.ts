// Node Shapes
import workloadImage from '../../assets/img/graph/pf/node.svg';
import appImage from '../../assets/img/graph/pf/app.svg';
import serviceImage from '../../assets/img/graph/pf/service.svg';
import operationImage from '../../assets/img/graph/pf/operation.svg';
import serviceEntryImage from '../../assets/img/graph/pf/service-entry.svg';
// Node Colors
import nodeColorHealthyImage from '../../assets/img/graph/pf/node-color-healthy.svg';
import nodeColorWarningImage from '../../assets/img/graph/pf/node-color-warning.svg';
import nodeColorDangerImage from '../../assets/img/graph/pf/node-color-danger.svg';
import nodeColorIdleImage from '../../assets/img/graph/pf/node-color-idle.svg';
// Node Background
import externalNamespaceImage from '../../assets/img/graph/external-namespace.svg';
import restrictedNamespaceImage from '../../assets/img/graph/restricted-namespace.svg';
// Edges
import edgeSuccessImage from '../../assets/img/graph/pf/edge-success.svg';
import edgeDangerImage from '../../assets/img/graph/pf/edge-danger.svg';
import edgeWarnImage from '../../assets/img/graph/pf/edge-warn.svg';
import edgeIdlemage from '../../assets/img/graph/pf/edge-idle.svg';
import edgeTcpImage from '../../assets/img/graph/pf/edge-tcp.svg';
import edgeMtlsImage from '../../assets/img/graph/mtls-badge.svg';
// Traffic Animation
import trafficHealthyImage from '../../assets/img/graph/pf/traffic-healthy-request.svg';
import trafficFailedImage from '../../assets/img/graph/pf/traffic-failed-request.svg';
import trafficTcpImage from '../../assets/img/graph/pf/traffic-tcp.svg';
// Badges
import badgeCircuitBreakerImage from '../../assets/img/graph/node-badge-circuit-breaker.svg';
import badgeFaultInjectionImage from '../../assets/img/graph/node-badge-fault-injection.svg';
import badgeGatewaysImage from '../../assets/img/graph/node-badge-gateways.svg';
import badgeMirroringImage from '../../assets/img/graph/node-badge-mirroring.svg';
import badgeMissingSidecarImage from '../../assets/img/graph/node-badge-missing-sidecar.svg';
import badgeRequestTimeoutImage from '../../assets/img/graph/node-badge-request-timeout.svg';
import badgeTrafficShiftingSourceImage from '../../assets/img/graph/node-badge-traffic-shifting.svg';
import badgeTrafficSourceImage from '../../assets/img/graph/node-badge-traffic-source.svg';
import badgeVirtualServicesImage from '../../assets/img/graph/node-badge-virtual-services.svg';
import badgeWorkloadEntryImage from '../../assets/img/graph/node-badge-workload-entry.svg';
import badgeWaypointImage from '../../assets/img/graph/node-badge-waypoint.svg';
import { t } from 'utils/I18nUtils';
import { serverConfig } from '../../config';

export interface GraphLegendItem {
  data: GraphLegendItemRow[];
  isBadge?: boolean;
  title: string;
}

export interface GraphLegendItemRow {
  icon: string;
  label: string;
}

export const legendData = (): GraphLegendItem[] => {
  const nodeBadges = [
    { label: t('Circuit Breaker'), icon: badgeCircuitBreakerImage },
    { label: t('Fault Injection'), icon: badgeFaultInjectionImage },
    { label: t('Gateway'), icon: badgeGatewaysImage },
    { label: t('Mirroring'), icon: badgeMirroringImage },
    { label: t('Missing Sidecar'), icon: badgeMissingSidecarImage },
    { label: t('Request Timeout'), icon: badgeRequestTimeoutImage },
    { label: t('Traffic Shifting / TCP Traffic Shifting'), icon: badgeTrafficShiftingSourceImage },
    { label: t('Traffic Source'), icon: badgeTrafficSourceImage },
    { label: t('Virtual Service / Request Routing'), icon: badgeVirtualServicesImage },
    { label: t('Workload Entry'), icon: badgeWorkloadEntryImage }
  ];

  if (serverConfig.ambientEnabled) {
    nodeBadges.push({ label: t('Waypoint'), icon: badgeWaypointImage });
  }

  return [
    {
      title: t('Node Shapes'),
      data: [
        { label: t('Workload'), icon: workloadImage },
        { label: t('App'), icon: appImage },
        { label: t('Operation'), icon: operationImage },
        { label: t('Service'), icon: serviceImage },
        { label: t('Service Entry'), icon: serviceEntryImage }
      ]
    },
    {
      title: t('Node Colors'),
      data: [
        { label: t('Healthy'), icon: nodeColorHealthyImage },
        { label: t('Warn'), icon: nodeColorWarningImage },
        { label: t('Unhealthy'), icon: nodeColorDangerImage },
        { label: t('Idle'), icon: nodeColorIdleImage }
      ]
    },
    {
      title: t('Node Background'),
      data: [
        { label: t('Unselected Namespace'), icon: externalNamespaceImage },
        { label: t('Restricted / External'), icon: restrictedNamespaceImage }
      ]
    },
    {
      title: t('Edges'),
      data: [
        { label: t('Failure'), icon: edgeDangerImage },
        { label: t('Degraded'), icon: edgeWarnImage },
        { label: t('Healthy'), icon: edgeSuccessImage },
        { label: t('TCP Connection'), icon: edgeTcpImage },
        { label: t('Idle'), icon: edgeIdlemage },
        { label: t('mTLS (badge)'), icon: edgeMtlsImage }
      ]
    },
    {
      title: t('Traffic Animation'),
      data: [
        { label: t('Healthy Request'), icon: trafficHealthyImage },
        { label: t('Failed Request'), icon: trafficFailedImage },
        { label: t('TCP Traffic'), icon: trafficTcpImage }
      ]
    },
    {
      title: t('Node Badges'),
      isBadge: true,
      data: nodeBadges
    }
  ];
};
