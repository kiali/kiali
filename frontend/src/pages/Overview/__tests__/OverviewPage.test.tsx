import * as React from 'react';
import { MemoryRouter } from 'react-router-dom-v5-compat';
import { Provider } from 'react-redux';
import { shallowToJson } from 'enzyme-to-json';
import { mount, shallow, ReactWrapper } from 'enzyme';
import { OverviewPage, OverviewPageComponent } from '../OverviewPage';
import { FilterSelected } from '../../../components/Filters/StatefulFilters';
import * as API from '../../../services/Api';
import { AppHealth, NamespaceAppHealth, HEALTHY, FAILURE, DEGRADED } from '../../../types/Health';
import { store } from '../../../store/ConfigStore';
import { MTLSStatuses } from '../../../types/TLSStatus';
import { FilterType, ActiveFiltersInfo } from 'types/Filters';
import { healthFilter } from 'components/Filters/CommonFilters';
import { nameFilter } from '../Filters';
import { DEFAULT_LABEL_OPERATION } from '../../../types/Filters';
import { CLUSTER_DEFAULT } from '../../../types/Graph';

const mockAPIToPromise = (func: keyof typeof API, obj: any, encapsData: boolean): Promise<void> => {
  return new Promise((resolve, reject) => {
    jest.spyOn(API, func).mockImplementation(() => {
      return new Promise(r => {
        if (encapsData) {
          r({ data: obj });
        } else {
          r(obj);
        }
        setTimeout(() => {
          try {
            resolve();
          } catch (e) {
            reject(e);
          }
        }, 2);
      });
    });
  });
};

const mockNamespaces = (names: string[]): Promise<void> => {
  return mockAPIToPromise(
    'getNamespaces',
    names.map(n => ({ name: n, cluster: CLUSTER_DEFAULT })),
    true
  );
};

const mockNamespaceHealth = (namespaces: string[], obj: NamespaceAppHealth): Promise<void> => {
  const res: Map<string, NamespaceAppHealth> = new Map<string, NamespaceAppHealth>();
  namespaces.forEach(namespace => {
    res[namespace] = obj;
  });
  return mockAPIToPromise('getClustersAppHealth', res, false);
};

let mounted: ReactWrapper<any, any> | null;

const mountPage = (): void => {
  mounted = mount(
    <Provider store={store}>
      <MemoryRouter>
        <OverviewPage />
      </MemoryRouter>
    </Provider>
  );
};

const genActiveFilters = (filter: FilterType, values: string[]): ActiveFiltersInfo => {
  return {
    filters: values.map(v => {
      return {
        category: filter.category,
        value: v
      };
    }),
    op: 'or'
  };
};

const concat = (f1: ActiveFiltersInfo, f2: ActiveFiltersInfo): ActiveFiltersInfo => {
  return {
    filters: f1.filters.concat(f2.filters),
    op: f1.op
  };
};

describe('Overview page', () => {
  beforeEach(() => {
    mounted = null;

    // Ignore other calls
    mockAPIToPromise('getClustersServiceHealth', null, false);
    mockAPIToPromise('getClustersWorkloadHealth', null, false);
    mockAPIToPromise('getClustersMetrics', null, false);
    mockAPIToPromise('getNamespaceMetrics', null, false);
    mockAPIToPromise('getNamespaceTls', null, false);
    mockAPIToPromise('getConfigValidations', null, false);
    mockAPIToPromise('getAllIstioConfigs', null, false);
    mockAPIToPromise('getIstioPermissions', {}, false);
  });

  afterEach(() => {
    jest.clearAllMocks();

    if (mounted) {
      mounted.unmount();
    }
  });

  it('renders initial layout', () => {
    const wrapper = shallow(
      <OverviewPageComponent
        lastRefreshAt={1720526431902}
        meshStatus={MTLSStatuses.NOT_ENABLED}
        navCollapse={false}
        duration={600}
        refreshInterval={10000}
        language={'en'}
        kiosk={''}
        minTLS={''}
        istioAPIEnabled={false}
      />
    );
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });

  it('renders all without filters', done => {
    FilterSelected.setSelected({ filters: [], op: DEFAULT_LABEL_OPERATION });
    Promise.all([
      mockNamespaces(['a', 'b', 'c']),
      mockNamespaceHealth(['a', 'b', 'c'], {
        app1: {
          getGlobalStatus: () => HEALTHY
        } as AppHealth,
        app2: {
          getGlobalStatus: () => FAILURE
        } as AppHealth
      })
    ]).then(() => {
      mounted!.update();
      // All 3 namespaces rendered
      expect(mounted!.find('Card')).toHaveLength(3);
      done();
    });
    mountPage();
  });

  it('filters failures match', done => {
    FilterSelected.setSelected(genActiveFilters(healthFilter, ['Failure']));
    Promise.all([
      mockNamespaces(['a']),
      mockNamespaceHealth(['a'], {
        app1: {
          getGlobalStatus: () => DEGRADED
        } as AppHealth,
        app2: {
          getGlobalStatus: () => FAILURE
        } as AppHealth,
        app3: {
          getGlobalStatus: () => HEALTHY
        } as AppHealth
      })
    ]).then(() => {
      mounted!.update();
      expect(mounted!.find('Card')).toHaveLength(1);
      done();
    });
    mountPage();
  });

  it('filters failures no match', done => {
    FilterSelected.setSelected(genActiveFilters(healthFilter, ['Failure']));
    Promise.all([
      mockNamespaces(['a']),
      mockNamespaceHealth(['a'], {
        app1: {
          getGlobalStatus: () => DEGRADED
        } as AppHealth,
        app2: {
          getGlobalStatus: () => HEALTHY
        } as AppHealth,
        app3: {
          getGlobalStatus: () => HEALTHY
        } as AppHealth
      })
    ]).then(() => {
      mounted!.update();
      expect(mounted!.find('Card')).toHaveLength(0);
      done();
    });
    mountPage();
  });

  it('multi-filters health match', done => {
    FilterSelected.setSelected(genActiveFilters(healthFilter, ['Failure', 'Degraded']));
    Promise.all([
      mockNamespaces(['a']),
      mockNamespaceHealth(['a'], {
        app1: {
          getGlobalStatus: () => DEGRADED
        } as AppHealth,
        app2: {
          getGlobalStatus: () => HEALTHY
        } as AppHealth
      })
    ]).then(() => {
      mounted!.update();
      expect(mounted!.find('Card')).toHaveLength(1);
      done();
    });
    mountPage();
  });

  it('multi-filters health no match', done => {
    FilterSelected.setSelected(genActiveFilters(healthFilter, ['Failure', 'Degraded']));
    Promise.all([
      mockNamespaces(['a']),
      mockNamespaceHealth(['a'], {
        app1: {
          getGlobalStatus: () => HEALTHY
        } as AppHealth,
        app2: {
          getGlobalStatus: () => HEALTHY
        } as AppHealth
      })
    ]).then(() => {
      mounted!.update();
      expect(mounted!.find('Card')).toHaveLength(0);
      done();
    });
    mountPage();
  });

  it('filters namespaces info name match', done => {
    FilterSelected.setSelected(genActiveFilters(nameFilter, ['bc']));
    Promise.all([
      mockNamespaces(['abc', 'bce', 'ced']),
      mockNamespaceHealth(['abc', 'bce', 'ced'], {
        app1: {
          getGlobalStatus: () => HEALTHY
        } as AppHealth
      })
    ]).then(() => {
      mounted!.update();
      expect(mounted!.find('Card')).toHaveLength(2);
      done();
    });
    mountPage();
  });

  it('filters namespaces info name no match', done => {
    FilterSelected.setSelected(genActiveFilters(nameFilter, ['yz']));
    mockNamespaces(['abc', 'bce', 'ced']).then(() => {
      mounted!.update();
      expect(mounted!.find('Card')).toHaveLength(0);
      done();
    });
    mountPage();
  });

  it('filters namespaces info name and health match', done => {
    FilterSelected.setSelected(
      concat(genActiveFilters(nameFilter, ['bc']), genActiveFilters(healthFilter, ['Healthy']))
    );
    Promise.all([
      mockNamespaces(['abc', 'bce', 'ced']),
      mockNamespaceHealth(['abc', 'bce', 'ced'], {
        app1: {
          getGlobalStatus: () => HEALTHY
        } as AppHealth
      })
    ]).then(() => {
      mounted!.update();
      expect(mounted!.find('Card')).toHaveLength(2);
      done();
    });
    mountPage();
  });

  it('filters namespaces info name and health no match', done => {
    FilterSelected.setSelected(
      concat(genActiveFilters(nameFilter, ['bc']), genActiveFilters(healthFilter, ['Healthy']))
    );
    Promise.all([
      mockNamespaces(['abc', 'bce', 'ced']),
      mockNamespaceHealth(['abc', 'bce', 'ced'], {
        app1: {
          getGlobalStatus: () => DEGRADED
        } as AppHealth
      })
    ]).then(() => {
      mounted!.update();
      expect(mounted!.find('Card')).toHaveLength(0);
      done();
    });
    mountPage();
  });
});
