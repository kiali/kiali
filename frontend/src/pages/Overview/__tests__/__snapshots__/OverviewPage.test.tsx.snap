// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Overview page renders initial layout 1`] = `
<Fragment>
  <Connect(OverviewToolbarComponent)
    displayMode={1}
    onChange={[Function]}
    onError={[Function]}
    onRefresh={[Function]}
    setDisplayMode={[Function]}
    sort={[Function]}
    statefulFilterRef={
      Object {
        "current": null,
      }
    }
  />
  <EmptyOverview
    filteredNamespaces={Array []}
    loaded={false}
    refreshInterval={10000}
  >
    <RenderComponentScroll
      className="kiali_f10gijvm"
    >
      <Grid />
    </RenderComponentScroll>
  </EmptyOverview>
  <OverviewTrafficPolicies
    duration={600}
    hideConfirmModal={[Function]}
    isOpen={false}
    kind=""
    load={[Function]}
    nsTarget=""
    opTarget=""
  />
</Fragment>
`;
